### ISSUE TITLE
Easy Collaboration with Shared Cluster

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#2414]
Background from #2413 

The hard part of this is deciding on how we want to recommend how to develop w pachyderm. A lot of the advantages unique to pachyderm really shine when you're using real data sets. So collaboration on a shared cluster that's hooked into real data is a use case we want to make sure is very well supported.

I'm not sure how we can (or if we should) integrate this proposal w the local development proposal.

## Goal

Enable collaborative development on pipelines on a shared cluster. Specifically:

1) As a user, I can rapidly iterate on my code in a shared cluster
2) As a user, I can easily collaborate on my code in a shared cluster
3) As a user, I can easily deploy my code to a production environment

## Proposal

The meat of this is really some sugar pachyderm can provide plus some conventions we require from users. Basically we're going to take what we did for the client in #2413 and make it as easy as possible to do with pachyderm. The challenge is allowing a single pipeline spec to accommodate multiple users / development, without generating a lot of noise (in the output of `list-job`)

Integration:

- Add a GH integration (similar to travis) that allows Pachyderm to get triggered on a post-commit hook
- Add a `dev-branches` field on the pipeline spec
- When a new commit comes in on one of the `dev-branches`, re-run the pipeline
    - (probably w `re-process`? though that could be configurable ... perhaps you only want to re-run the last 10 commits to get a sample of the real data but not have to re-run the whole data set)
    - Whenever a pipeline is run in this fashion, we populate the env variable `PACHYDERM_DEV_BRANCH`
    - We'd need their pipeline transform code to check if this value was populated, and if so, after cloning the git repo, do an additional `git checkout $PACHYDERM_DEV_BRANCH`
- The output of `list-job` would only include jobs from non-dev branches
    - To see the output for your branch, you'd need to provide an additional `--dev-branch` field

This satisfies goals # 1 and # 2. To satisfy # 3, additionally:

- We add a `tag` field to the pipeline
- When this value is set, it is provided as an environment variable to the container, e.g. `PACHYDERM_PIPELINE_TAG`
- So that their pipeline transform can use it
- Note: We'd need their code, by convention, to prefer using the value of `PACHYDERM_PIPELINE_TAG` if present ... over checking out a specific branch (e.g. the presence of the `PACHYDERM_DEV_BRANCH` variable)
    - e.g. `git checkout $PACHYDERM_PIPELINE_TAG`
- This allows them to deploy a pipeline to a production cluster, and when doing so, locking that code to a particular version (that if tagged ... presumable has been tested and released) ... implicitly encouraging good code / testing habits

UX:

It's worth restating how this interface would work for a user. It's pretty simple:

- Create a docker file that just contains the runtime dependencies
    - This is done once, and is 'set and forget'
    - So that most of the time a Data Scientist spends is spent coding, not with docker
- Create a pipeline transform that git clones a repo and checks out a tag or branch name
- Add the 'pachyderm' integration to their GH repository
- Create a pipeline spec that lists out the branches e.g. `["alice", "steve", "dev"]` that it should listen to
- When ready to deploy a pipeline (to their long running dev cluster's 'main' pipeline ... or to an entirely different production cluster), they add/update the `tag` field of a pipeline w the GH tag of the tested + released version

## Shortcomings

- It feels weird to have their pipeline cloning code in production
- But this allows increased similarity between dev and production environments
- This also supplants the need for private registries ... which we have mixed feelings about ... and different clients will have different (strong!) opinions about as well
- The two new environment variables are a bit redundant ... we could populate a single one w The Right Value but this may be a bit confusing to users

## Advantages

- I like that this enables some strong conventions for users. Often our users need guidance around dev workflows and deployment.
- I like that it makes it easy to collaborate on a Real Cluster with Real Data ... this is something that pachyderm is uniquely good for, and its important that we enable users to do this as easily as possible

## Brainstorming

Some other ideas not chosen / rough ideas.

1) file watcher + scp

- What if we had a local file watcher ... which scp'd your code changes to all the cloud nodes on changes?
- This way you could still use the host mount trick ... though you'd easily step on each others toes using the same cluster ... so no that won't work

2) using a 'code' repo

- so that we just put their local code to this repo ... which is then already mounted in to the pipeline
- this has some of the advantages of the scp approach ... but naturally lets you have branches for concurrent development
- but feels a bit weird ... we're not trying to be git

**Comment by sjezewski** [comment]
Something to consider in the new model:

https://github.com/pachyderm/pachyderm/issues/2412

We may want to be able to 'tag' the job or something w the git hash that was checked out to be run.

**Comment by jdoliner** [comment]
Filling in some details on how I think we could accomplish this.
I think we should add a new type of input called a `git` input. It would look something like this:

```json
{
    "name": "pachyderm",
    "url": "github.com/pachyderm/pachyderm",
    "branch": "master"
}
```

This would be an input that would yield a single datum which would be the whole repo. Having this datum in a pipeline would cause the pipeline to subscribe to the repo via commit hook so that when a new commit comes in on the specified branch the pipeline gets triggered. This would allow the classic Heroku style workflow of `git push` as a way to deploy your code to Pachyderm.

**Comment by sjezewski** [comment]
Love it. Let's do it.

As an implementation detail, I think the easiest way to enable users to do this is as a github integration (a la travis), as opposed to requiring that they specify a GH ssh key for a machine user (which requires a fair amount of setup).

**Comment by sjezewski** [comment]
Some discussion on implementation method / iteration:

## Iteration 0

### Goal

As a user, I can specify a github repository as a pipeline input, that gets triggered by new pushes (NOT every new commit) to that repository.

As a user, I should only have to add some config to my GH repo for integration ... I should not have to create a new GH machine user / SSH key.

As a user, this service should work for public or private repositories.

### Implementation

Note: This will only be supported on cloud based deployments

- When a user specifies a GH repo as an input we:
    - Provide a URL for them to copy/paste into their repo's config for a GH integration
    - Create a pachyderm repo w the GH repo name
    - This repo will have a field that marks it as a GH repository
- Create a github service on the pachd node that accepts a POST
    - This is the integration point w GH's webhook API so that we're notified on new commits being pushed
    - We'll use the `push` event
- Upon receiving a new push notification, we'll
    - Add a new commit to that repository containing the commit SHA of `HEAD` (the new commit we will want to clone)
- Then, when a worker processes a datum that contains an input repo that is marked as a GH repo:
    - Instead of mounting the files (in this case just the commit SHA), instead we ...
    - Clone the contents of that GH repo at the commit SHA provided on the latest commit 
    - Then, we proceed processing the job as normal

### Open Questions

- Exactly how we can integrate w GH to authenticate to enable cloning private repos (should be do-able ... travis does this)

## Iteration 1

Nice to haves:

- Specify a branch name as well as a repository
- Make our pipeline creation smart enough to automatically provision a cloud load balancer + DNS / etc so that we can provide the user w a public domain name to copy/paste into GH
- Update the UI to support this feature during pipeline creation
- Update UI to link back to the GH repo at the SHA in question when navigating to a GH input type

**Comment by sjezewski** [comment]
Note ... if we do implement that 'load balancer / DNS / firewall' nice to have ... it'll also buy us a proper URL to use for the dashboard (no more port forwarding).

**Comment by sjezewski** [comment]
Breadcrumb for myself here ... [this project](https://github.com/kubernetes-incubator/external-dns) looks to have what we need in terms of an external DNS setup. Right now (v0.4) it almost has kops support too, which would be great (so we don't have to do our DNS hacking on our aws.sh deploy).

**Comment by sjezewski** [comment]
And this is what we have today via k8s 1.7 [to expose a public IP address](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/)

**Comment by sjezewski** [comment]
Doing some more reading on the GH integration options (a great [tradeoff writeup is here](https://developer.github.com/apps/building-integrations/setting-up-a-new-integration/about-choosing-an-integration-type/) ) I'm not entirely clear which option we want.

On the one hand a GH app gets clone access right away ... on the other hand ... it would be nice to have any user click a button to enable this support. However, then I'd have to deal with / hook in oauth integration (and we'd need to support that on the UI as well).

So for now to keep things simple ... the first iteration will only include the ability to clone from a public repo. We'll just use the webhook feature directly.

**Comment by sjezewski** [comment]
Just an update on what we have merged so far that'll make the 1.7 release:

- we support git inputs (specify a repo url / branch)
- we don't yet support cloning at a SHA (which means we clone then checkout ... which incurs overhead for large repos)
- we don't yet support private GH repositories

**Comment by sjezewski** [comment]
I've described what will land for 1.7

Beyond that, we'll create separate issues for further improvments.
